File: common/src/main/java/com/ssblur/scriptor/word/subject/RuneSubject.java
package com.ssblur.scriptor.word.subject;

import com.ssblur.scriptor.block.ScriptorBlocks;
import com.ssblur.scriptor.blockentity.RuneBlockEntity;
import com.ssblur.scriptor.helpers.targetable.EntityTargetable;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import com.ssblur.scriptor.messages.TraceNetwork;
import com.ssblur.scriptor.word.Spell;
import com.ssblur.scriptor.word.descriptor.visual.ColorDescriptor;
import com.ssblur.scriptor.word.descriptor.Descriptor;
import net.minecraft.core.BlockPos;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class RuneSubject extends Subject implements InventorySubject{

  @Override
  public CompletableFuture<List<Targetable>> getTargets(Targetable caster, Spell spell) {
    var result = new CompletableFuture<List<Targetable>>();
    if(caster instanceof EntityTargetable entityTargetable && entityTargetable.getTargetEntity() instanceof Player player) {
      TraceNetwork.requestTraceData(player, target -> {

        int color = 0xa020f0;
        for(Descriptor d: spell.descriptors())
          if(d instanceof ColorDescriptor descriptor)
            color = descriptor.getColor();

        BlockPos pos = target.getTargetBlockPos();
        Level level = caster.getLevel();

        if(!level.getBlockState(pos).getMaterial().isReplaceable())
          return;

        level.setBlockAndUpdate(pos, ScriptorBlocks.RUNE.get().defaultBlockState());
        var entity = level.getBlockEntity(pos);
        if(entity instanceof RuneBlockEntity runeBlockEntity) {
          runeBlockEntity.owner = player;
          runeBlockEntity.future = result;
          runeBlockEntity.spell = spell;
          runeBlockEntity.color = color;
          runeBlockEntity.setChanged();
        }
      });
    } else {
      int color = 0xa020f0;
      for(Descriptor d: spell.descriptors())
        if(d instanceof ColorDescriptor descriptor)
          color = descriptor.getColor();

      BlockPos pos = caster.getTargetBlockPos();
      Level level = caster.getLevel();

      if(!level.getBlockState(pos).getMaterial().isReplaceable()) {
        result.complete(List.of());
        return result;
      }

      level.setBlockAndUpdate(pos, ScriptorBlocks.RUNE.get().defaultBlockState());
      var entity = level.getBlockEntity(pos);
      if(entity instanceof RuneBlockEntity runeBlockEntity) {
        runeBlockEntity.future = result;
        runeBlockEntity.spell = spell;
        runeBlockEntity.color = color;
        runeBlockEntity.setChanged();
      }
    }
    return result;
  }

  @Override
  public Cost cost() { return new Cost(1, COSTTYPE.ADDITIVE); }

  @Override
  public void castOnItem(Spell spell, Player player, ItemStack slot) {
    // add a one-time cast enchant
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/word/subject/RuneSubject.kt
package com.ssblur.scriptor.word.subject

import com.ssblur.scriptor.api.word.Subject
import com.ssblur.scriptor.block.ScriptorBlocks
import com.ssblur.scriptor.blockentity.RuneBlockEntity
import com.ssblur.scriptor.color.CustomColors.getColor
import com.ssblur.scriptor.helpers.targetable.EntityTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.network.server.TraceNetwork
import com.ssblur.scriptor.word.Spell
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import java.util.concurrent.CompletableFuture

class RuneSubject : Subject(), InventorySubject {
    override fun getTargets(caster: Targetable, spell: Spell): CompletableFuture<List<Targetable>> {
        val result = CompletableFuture<List<Targetable>>()
        if (caster is EntityTargetable && caster.targetEntity is Player) {
            val player = caster.targetEntity as Player
            TraceNetwork.requestTraceData(player) { target: Targetable ->
                val color = getColor(spell.deduplicatedDescriptorsForSubjects())
                val pos = target.targetBlockPos
                val level = caster.level

                if (!level.getBlockState(pos).canBeReplaced()) return@requestTraceData

                level.setBlockAndUpdate(pos, ScriptorBlocks.RUNE.get().defaultBlockState())
                val entity = level.getBlockEntity(pos)
                if (entity is RuneBlockEntity) {
                    entity.owner = player
                    entity.future = result
                    entity.spell = spell
                    entity.color = color
                    entity.setChanged()
                }
            }
        } else {
            val color = getColor(spell.deduplicatedDescriptorsForSubjects())

            val pos = caster.targetBlockPos
            val level = caster.level

            if (!level.getBlockState(pos).canBeReplaced()) {
                result.complete(listOf())
                return result
            }

            level.setBlockAndUpdate(pos, ScriptorBlocks.RUNE.get().defaultBlockState())
            val entity = level.getBlockEntity(pos)
            if (entity is RuneBlockEntity) {
                entity.future = result
                entity.spell = spell
                entity.color = color
                entity.setChanged()
            }
        }
        return result
    }

    override fun cost(): Cost {
        return Cost(1.0, COSTTYPE.ADDITIVE)
    }

    override fun castOnItem(spell: Spell, player: Player, slot: ItemStack) {
        // add a one-time cast
    }
}
